docs/seed.1
'\" t
.\"     Title: seed
.\"    Author: Jonathan Meyer, Mike Holt, Derick Faller, John Tobe
.\" Generator: Asciidoctor 1.5.7.1
.\"      Date: 2018-06-11
.\"    Manual: SEED
.\"    Source: SEED
.\"  Language: English
.\"
.TH "SEED" "1" "2018-06-11" "SEED" "SEED"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
seed \- A test runner for Seed spec compliant algorithms
.SH "SYNOPSIS"
.sp
\fBseed\fP [COMMAND] [OPTIONS]
.SH "DESCRIPTION"
.sp
The Seed team provides a fully featured Command\-Line Interface (CLI) for algorithm developers looking to offer Seed compliant packaging of their jobs. Using the CLI will enable you to quickly iterate without getting bogged down learning how to interface directly with the underlying container technologies (Docker). It also simplifies the process of attaching Seed metadata to your algorithm prior to publish.
.SH "COMMANDS"
.sp
Seed CLI offers sub\-commands that correlate to common operations performed by an algorithm developer. These can be seen from the command\-line by launching the CLI without any arguments.
.SS "build"
.sp
Builds Seed compliant Docker image
.sp
seed build [\-d JOB_DIRECTORY] [\-u USER_NAME \-p PASSWORD]
.sp
\fB\-d, \-directory\fP
.RS 4
Directory containing Seed spec and Dockerfile (default is current directory).
.RE
.sp
\fB\-u, \-user\fP
.RS 4
Username to login if needed to pull images (default anonymous).
.RE
.sp
\fB\-p, \-password\fP
.RS 4
Password to login if needed to pull images (default anonymous).
.RE
.SS "batch"
.sp
Executes Seed compliant Docker image over multiple iterations
.sp
seed batch \-in IMAGE_NAME [\-b BATCH_FILE | \-d BATCH_DIRECTORY] [\-e SETTING=SETTING_VALUE] [\-m MOUNT_KEY=HOST_PATH] [\-o OUTPUT_DIRECTORY]
.sp
\fB\-in, \-imageName\fP
.RS 4
Docker image name to run; Required
.RE
.sp
\fB\-b,  \-batch\fP
.RS 4
Optional file specifying input keys and file mapping for batch processing. Supersedes directory flag.
.RE
.sp
\fB\-d, \-directory\fP
.RS 4
Alternative to batch file; Specifies a directory of files to batch process (default is current directory)
.RE
.sp
\fB\-e,  \-setting\fP
.RS 4
Specifies the key/value setting values of the seed spec in the format SETTING_KEY=VALUE
.RE
.sp
\fB\-m, \-mount\fP
.RS 4
Specifies the key/value mount values of the seed spec in the format MOUNT_KEY=HOST_PATH
.RE
.sp
\fB\-o, \-outDir\fP
.RS 4
Job Output Directory Location
.RE
.sp
\fB\-rm\fP
.RS 4
Automatically remove the container when it exits (docker run \-\-rm)
.RE
.sp
\fB\-s,  \-schema\fP
.RS 4
External Seed metadata schema file; Overrides built in schema to validate side\-car metadata files
.RE
.SS "init"
.sp
Initialize new project with example seed.manifest.json file
.sp
seed init [\-d JOB_DIRECTORY]
.sp
\fB\-d  \-directory\fP
.RS 4
Directory to place seed.manifest.json example. (default is current directory)
.RE
.SS "list"
.sp
Allows for listing of all Seed compliant images residing on the local system
.sp
seed list
.SS "publish"
.sp
Publishes Seed compliant images to remote Docker registry
.sp
seed publish \-in IMAGE_NAME [\-r REGISTRY_NAME] [\-o ORG_NAME] [\-u username] [\-p password] [Conflict Options]
.sp
\fB\-in, \-imageName\fP
.RS 4
Docker image name to publish
.RE
.sp
\fB\-r, \-registry\fP
.RS 4
Specifies a specific registry to publish the image
.RE
.sp
\fB\-o, \-org\fP
.RS 4
Specifies a specific organization to publish the image
.RE
.sp
\fB\-u, \-user\fP
.RS 4
Username to login if needed to publish images (default anonymous).
.RE
.sp
\fB\-p, \-password\fP
.RS 4
Password to login if needed to publish images (default anonymous).
.RE
.sp
\fB\-f\fP
.RS 4
Overwrite remote image if publish conflict found
.RE
.sp
\fBCONFLICT OPTIONS\fP
.RS 4
If the force flag (\-f) is not set, the following options specify how a publish conflict is handled:
    seed publish ... \-f [\-d SEED_DIRECTORY] [\-pp] [\-pm] [\-P] [\-jp] [\-J]
.RE
.sp
\fB\-d, \-directory\fP
.RS 4
Specifies the directory containing the seed.manifest.json and dockerfile to rebuild the image.
.RE
.sp
\fB\-pp\fP
.RS 4
Force Patch version bump of \(aqpackageVersion\(aq in manifest on disk if publish conflict found
.RE
.sp
\fB\-pm\fP
.RS 4
Force Minor version bump of \(aqpackageVersion\(aq in manifest on disk if publish conflict found
.RE
.sp
\fB\-P\fP
.RS 4
Force Major version bump of \(aqpackageVersion\(aq in manifest on disk if publish conflict found
.RE
.sp
\fB\-jp\fP
.RS 4
Force Patch version bump of \(aqjobVersion\(aq in manifest on disk if publish conflict found
.RE
.sp
\fB\-jm\fP
.RS 4
Force Minor version bump of \(aqjobVersion\(aq in manifest on disk if publish conflict found
.RE
.sp
\fB\-J\fP
.RS 4
Force Major version bump of \(aqjobVersion\(aq in manifest on disk if publish conflict found
.RE
.sp
\fBEXAMPLE\fP
    This will build a new image example\-0.2.0\-seed:1.0.0 and publish it to hub.docker.com/geoint
    seed publish \-in example\-0.1.3\-seed:0.1.3 \-r hub.docker.com \-o geoint \-jm \-P
.SS "pull"
.sp
Allows for pulling Seed compliant images from remote Docker registry
.sp
seed pull \-in IMAGE_NAME [\-r REGISTRY_NAME] [\-o ORGANIZATION_NAME] [\-u USER_NAME] [\-p PASSWORD]
.sp
\fB\-in, \-imageName\fP
.RS 4
Docker image name to pull
.RE
.sp
\fB\-r, \-registry\fP
.RS 4
Specifies a specific registry (default is index.docker.io).
.RE
.sp
\fB\-o, \-org\fP
.RS 4
Specifies a specific organization (default is no organization).
.RE
.sp
\fB\-u, \-user\fP
.RS 4
Username to login to remote registry (default anonymous).
.RE
.sp
\fB\-p, \-password\fP
.RS 4
Password to login to remote registry (default anonymous).
.RE
.SS "run"
.sp
Executes Seed compliant Docker docker image
.sp
seed run \-in IMAGE_NAME [\-rm] [\-q] [\-i INPUT_FILE_KEY=INPUT_FILE_VALUE] [\-e SETTING_KEY=SETTING_VALUE] [\-m MOUNT_KEY=HOST_PATH] [\-o OUTPUT_DIRECTORY] [\-rep 5] [\-s SCHEMA_FILE]
.sp
\fB\-in, \-imageName\fP
.RS 4
Docker image name to run
.RE
.sp
\fB\-i, \-inputs\fP
.RS 4
Specifies the key/value input data values of the seed spec in the format INPUT_FILE_KEY=INPUT_FILE_VALUE
.RE
.sp
\fB\-e, \-setting\fP
.RS 4
Specifies the key/value setting values of the seed spec in the format SETTING_KEY=VALUE
.RE
.sp
\fB\-m, \-mount\fP
.RS 4
Specifies the key/value mount values of the seed spec in the format MOUNT_KEY=HOST_PATH
.RE
.sp
\fB\-o, \-outDir\fP
.RS 4
Job Output Directory Location
.RE
.sp
\fB\-rm\fP
.RS 4
Automatically remove the container when it exits (docker run \-\-rm)
.RE
.sp
\fB\-q, \-quiet\fP
.RS 4
Suppress stdout when running docker image
.RE
.sp
\fB\-rep, \-repetitions\fP
.RS 4
Run docker image multiple times (i.e. \-rep 5 runs the image 5 times)
.RE
.sp
\fB\-s, \-schema\fP
.RS 4
External Seed metadata schema file; Overrides built in schema to validate side\-car metadata files
.RE
.SS "search"
.sp
Allows for discovery of Seed compliant images hosted within a Docker registry (default is docker.io)
.sp
seed search [\-r REGISTRY_NAME] [\-o ORGANIZATION_NAME] [\-f FILTER] [\-u Username] [\-p password]
.sp
\fB\-r, \-registry\fP
.RS 4
Specifies a specific registry to search (default is index.docker.io).
.RE
.sp
\fB\-o, \-org\fP
.RS 4
Specifies a specific organization to filter (default is no filter).
.RE
.sp
\fB\-f, \-filter\fP
.RS 4
Specifies a filter to apply (default is no filter).
.RE
.sp
\fB\-u, \-user\fP
.RS 4
Username to login to remote registry (default is anonymous).
.RE
.sp
\fB\-p, \-password\fP
.RS 4
Password to login to remote registry (default is anonymous).
.RE
.SS "validate"
.sp
Validates a Seed spec
.sp
seed validate [\-d MANIFEST_DIRECTORY] [\-s SCHEMA_FILE]
.sp
\fB\-d, \-directory\fP
.RS 4
Specifies the directory in which Seed is located (default is the current directory)
.RE
.sp
\fB\-s, \-schema\fP
.RS 4
External Seed schema file; Overrides built in schema to validate Seed spec against
.RE
.SS "version"
.sp
Prints the current version of Seed spec
.sp
seed version
.SH "AUTHORS"
.sp
Jonathan Meyer
.sp
Mike Holt
.sp
Derick Faller
.sp
John Tobe

objects/Seed.go
package objects

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"strings"

	"github.com/ngageoint/seed-cli/constants"
	"github.com/ngageoint/seed-cli/util"
)

type Image struct {
	Name     string
	Registry string
	Org      string
	Manifest string
}

//Seed represents a seed.manifest.json object.
type Seed struct {
	SeedVersion string `json:"seedVersion"`
	Job         Job    `json:"job"`
}

type Job struct {
	Name           string     `json:"name"`
	JobVersion     string     `json:"jobVersion"`
	PackageVersion string     `json:"packageVersion"`
	Title          string     `json:"title,omitempty"`
	Description    string     `json:"description,omitempty"`
	Tags           []string   `json:"tags,omitempty"`
	Maintainer     Maintainer `json:"maintainer"`
	Timeout        int        `json:"timeout,omitempty"`
	Interface      Interface  `json:"interface,omitempty"`
	Resources      Resources  `json:"resources,omitempty"`
	Errors         []ErrorMap `json:"errors,omitempty"`
}

type Maintainer struct {
	Name         string `json:"name"`
	Organization string `json:"organization,omitempty"`
	Email        string `json:"email"`
	Url          string `json:"url,omitempty"`
	Phone        string `json:"phone,omitempty"`
}

type Interface struct {
	Command  string    `json:"command"`
	Inputs   Inputs    `json:"inputs,omitempty"`
	Outputs  Outputs   `json:"outputs,omitempty"`
	Mounts   []Mount   `json:"mounts,omitempty"`
	Settings []Setting `json:"settings,omitempty"`
}

type Resources struct {
	Scalar []Scalar `json:"scalar"`
}

type Scalar struct {
	Name            string  `json:"name"`
	Value           float64 `json:"value"`
	InputMultiplier float64 `json:"inputMultiplier"`
}

type Inputs struct {
	Files []InFile `json:"files,ommitempty"`
	Json  []InJson `json:"json,omitempty"`
}

type InFile struct {
	Name       string   `json:"name"`
	MediaTypes []string `json:"mediaTypes"`
	Multiple   bool     `json:"multiple"`
	Partial    bool     `json:"partial"`
	Required   bool     `json:"required"`
}

func (o *InFile) UnmarshalJSON(b []byte) error {
	type xInFile InFile
	xo := &xInFile{Multiple: false, Partial: false, Required: true}
	if err := json.Unmarshal(b, xo); err != nil {
		return err
	}
	*o = InFile(*xo)
	return nil
}

type InJson struct {
	Name     string `json:"name"`
	Type     string `json:"type"`
	Required bool   `json:"required"`
}

func (o *InJson) UnmarshalJSON(b []byte) error {
	type xInJson InJson
	xo := &xInJson{Required: true}
	if err := json.Unmarshal(b, xo); err != nil {
		return err
	}
	*o = InJson(*xo)
	return nil
}

type Outputs struct {
	Files []OutFile `json:"files,omitempty"`
	JSON  []OutJson `json:"json,omitempty"`
}

type OutFile struct {
	Name      string `json:"name"`
	MediaType string `json:"mediaType"`
	Multiple  bool   `json:"multiple"`
	Pattern   string `json:"pattern"`
	Required  bool   `json:"required"`
}

func (o *OutFile) UnmarshalJSON(b []byte) error {
	type xOutFile OutFile
	xo := &xOutFile{Multiple: false, Required: true}
	if err := json.Unmarshal(b, xo); err != nil {
		return err
	}
	*o = OutFile(*xo)
	return nil
}

type OutJson struct {
	Name     string `json:"name"`
	Key      string `json:"key"`
	Type     string `json:"type"`
	Required bool   `json:"required"`
}

func (o *OutJson) UnmarshalJSON(b []byte) error {
	type xOutJson OutJson
	xo := &xOutJson{Required: true}
	if err := json.Unmarshal(b, xo); err != nil {
		return err
	}
	*o = OutJson(*xo)
	return nil
}

type Mount struct {
	Name string `json:"name"`
	Path string `json:"path"`
	Mode string `json:"mode"`
}

func (o *Mount) UnmarshalJSON(b []byte) error {
	type xMount Mount
	xo := &xMount{Mode: "ro"}
	if err := json.Unmarshal(b, xo); err != nil {
		return err
	}
	*o = Mount(*xo)
	return nil
}

type Setting struct {
	Name   string `json:"name"`
	Secret bool   `json:"secret"`
}

func (o *Setting) UnmarshalJSON(b []byte) error {
	type xSetting Setting
	xo := &xSetting{Secret: false}
	if err := json.Unmarshal(b, xo); err != nil {
		return err
	}
	*o = Setting(*xo)
	return nil
}

type ErrorMap struct {
	Code        int    `json:"code"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Category    string `json:"category"`
}

func (o *ErrorMap) UnmarshalJSON(b []byte) error {
	type xErrorMap ErrorMap
	xo := &xErrorMap{Category: "job"}
	if err := json.Unmarshal(b, xo); err != nil {
		return err
	}
	*o = ErrorMap(*xo)
	return nil
}

//GetManifestLabel returns the seed.manifest.json as LABEL
//  com.ngageoint.seed.manifest contents
func GetManifestLabel(seedFileName string) string {
	// read the seed.manifest.json into a string
	seedbytes, err := ioutil.ReadFile(seedFileName)
	if err != nil {
		util.PrintUtil("ERROR: Error reading %s. %s\n", seedFileName,
			err.Error())
		os.Exit(1)
	}
	var seedbuff bytes.Buffer
	json.Compact(&seedbuff, seedbytes)
	seedbytes, err = json.Marshal(seedbuff.String())
	if err != nil {
		util.PrintUtil("ERROR: Error marshalling seed manifest. %s\n",
			err.Error())
	}

	// Escape forward slashes and dollar signs
	seed := string(seedbytes)
	seed = strings.Replace(seed, "$", "\\$", -1)
	seed = strings.Replace(seed, "/", "\\/", -1)

	return seed
}

//SeedFromImageLabel returns seed parsed from the docker image LABEL
func SeedFromImageLabel(imageName string) Seed {
	cmdStr := "inspect -f '{{index .Config.Labels \"com.ngageoint.seed.manifest\"}}'" + imageName
	util.PrintUtil(
		"INFO: Retrieving seed manifest from %s LABEL=com.ngageoint.seed.manifest\n",
		imageName)

	inspectCommand := exec.Command("docker", "inspect", "-f",
		"'{{index .Config.Labels \"com.ngageoint.seed.manifest\"}}'", imageName)

	errPipe, errr := inspectCommand.StderrPipe()
	if errr != nil {
		util.PrintUtil(
			"ERROR: error attaching to docker inspect command stderr. %s\n",
			errr.Error())
	}

	// Attach stdout pipe
	outPipe, errr := inspectCommand.StdoutPipe()
	if errr != nil {
		util.PrintUtil(
			"ERROR: error attaching to docker inspect command stdout. %s\n",
			errr.Error())
	}

	// Run docker inspect
	if err := inspectCommand.Start(); err != nil {
		util.PrintUtil("ERROR: error executing docker %s. %s\n", cmdStr,
			err.Error())
	}

	// Print out any std out
	seedBytes, err := ioutil.ReadAll(outPipe)
	if err != nil {
		fmt.Fprintf(os.Stdout, "ERROR: Error retrieving docker %s stdout.\n%s\n",
			cmdStr, err.Error())
	}

	// check for errors on stderr
	slurperr, _ := ioutil.ReadAll(errPipe)
	if string(slurperr) != "" {
		util.PrintUtil("ERROR: Error executing docker %s:\n%s\n",
			cmdStr, string(slurperr))
		util.PrintUtil("Exiting seed...\n")
		os.Exit(1)
	}

	// un-escape special characters
	seedStr := string(seedBytes)
	seedStr = strings.Replace(seedStr, "\\\"", "\"", -1)
	seedStr = strings.Replace(seedStr, "\\\"", "\"", -1) //extra replace to fix extra back slashes added by docker build command
	seedStr = strings.Replace(seedStr, "\\$", "$", -1)
	seedStr = strings.Replace(seedStr, "\\/", "/", -1)
	seedStr = strings.TrimSpace(seedStr)
	seedStr = strings.TrimSuffix(strings.TrimPrefix(seedStr, "'\""), "\"'")

	seed := &Seed{}

	err = json.Unmarshal([]byte(seedStr), &seed)
	if err != nil {
		util.PrintUtil("ERROR: Error unmarshalling seed: %s\n", err.Error())
	}

	return *seed
}

//SeedFromManifestFile returns seed struct parsed from seed file
func SeedFromManifestFile(seedFileName string) Seed {

	// Open and parse seed file into struct
	seedFile, err := os.Open(seedFileName)
	if err != nil {
		util.PrintUtil("ERROR: Error opening %s. Error received is: %s\n",
			seedFileName, err.Error())
		util.PrintUtil("Exiting seed...\n")
		os.Exit(1)
	}
	jsonParser := json.NewDecoder(seedFile)
	var seed Seed
	if err = jsonParser.Decode(&seed); err != nil {
		util.PrintUtil(
			"ERROR: A valid %s must be present in the working directory. Error parsing %s.\nError received is: %s\n",
			constants.SeedFileName, seedFileName, err.Error())
		util.PrintUtil("Exiting seed...\n")
		os.Exit(1)
	}

	return seed
}

//BuildImageName extracts the Docker Image name from the seed.json
// 	jobName-jobVersion-seed:pkgVersion
func BuildImageName(seed *Seed) string {
	var buffer bytes.Buffer

	buffer.WriteString(seed.Job.Name)
	buffer.WriteString("-")
	buffer.WriteString(seed.Job.JobVersion)
	buffer.WriteString("-seed")
	buffer.WriteString(":")
	buffer.WriteString(seed.Job.PackageVersion)

	return buffer.String()
}
